File: frameworks/dotnet/fastendpoints/UnitTests/EtagTests.cs
namespace UnitTests;

public sealed class EtagTests(App app) : TestBase<App>
{
    // rb:test etag.small,etag.large
    [Theory]
    [Trait("corpus", "etag.small")]
    [Trait("corpus", "etag.large")]
    [InlineData("small")]
    [InlineData("large")]
    public async Task The_answer_carries_a_validator(string size)
    {
        using HttpResponseMessage response = await app.Client.GetAsync($"/etag/{size}", Cancellation);

        Assert.NotNull(response.Headers.ETag);
        await Answer.IsAsync(Expected.Json($"items.{size}.json"), response);
    }

    // rb:test etag.match_large
    [Fact]
    [Trait("corpus", "etag.match_large")]
    public async Task The_validator_it_issued_is_answered_304_with_no_body()
    {
        using HttpResponseMessage first = await app.Client.GetAsync("/etag/large", Cancellation);

        using HttpResponseMessage second = await ConditionalAsync(first.Headers.ETag!.Tag);

        Assert.Equal(HttpStatusCode.NotModified, second.StatusCode);
        Assert.Empty(await Answer.BytesAsync(second));
    }

    // rb:test etag.stale_large
    [Fact]
    [Trait("corpus", "etag.stale_large")]
    public async Task A_validator_it_never_issued_is_answered_in_full()
    {
        using HttpResponseMessage response = await ConditionalAsync((string)Expected.Settings["staleEtag"]!);

        Assert.Equal(HttpStatusCode.OK, response.StatusCode);
        await Answer.IsAsync(Expected.Json("items.large.json"), response);
    }

    private Task<HttpResponseMessage> ConditionalAsync(string tag)
    {
        HttpRequestMessage request = new(HttpMethod.Get, "/etag/large");
        request.Headers.TryAddWithoutValidation("if-none-match", tag);
        return app.Client.SendAsync(request, Cancellation);
    }
}

File: frameworks/java/vertx/UnitTests/src/test/java/unittests/EtagTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.http.HttpResponse;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class EtagTests extends VertxApp {

    // rb:test etag.small,etag.large
    @ParameterizedTest
    @Tag("etag.small")
    @Tag("etag.large")
    @ValueSource(strings = {"small", "large"})
    void theAnswerCarriesAValidatorAndTheHandlerRunsEachTime(String size) throws Exception {
        HttpResponse<byte[]> first = get("/etag/" + size);
        HttpResponse<byte[]> second = get("/etag/" + size);

        assertNotNull(Answer.header(second, "etag"));
        Answer.is(Expected.json("items." + size + ".json"), second);
        assertTrue(Answer.serial(second) > Answer.serial(first));
    }

    // rb:test etag.match_large
    @Test
    @Tag("etag.match_large")
    void theValidatorItIssuedIsAnswered304WithNoBody() throws Exception {
        String tag = Answer.header(get("/etag/large"), "etag");

        HttpResponse<byte[]> response = get("/etag/large", "if-none-match", tag);

        assertEquals(304, response.statusCode());
        assertEquals(0, response.body().length);
    }

    // rb:test etag.stale_large
    @Test
    @Tag("etag.stale_large")
    void aValidatorItNeverIssuedIsAnsweredInFull() throws Exception {
        HttpResponse<byte[]> response = get("/etag/large", "if-none-match", Expected.settings().getString("staleEtag"));

        assertEquals(200, response.statusCode());
        Answer.is(Expected.json("items.large.json"), response);
    }
}
